pkg/trace/pb/decoder.go

```diff
@@ -38,6 +42,40 @@ func parseString(dc *msgp.Reader) (string, error) {
 	}
 }
 
+// parseStringDict
+func parseStringDict(dc *msgp.Reader, dict []string) (string, error) {
+	// read the generic representation type without decoding
+	t, err := dc.NextType()
+	if err != nil {
+		return "", err
+	}
+	switch t {
+	case msgp.BinType:
```

Comment:
I'm confused by the need for BinType and StrType cases here.
I'm guessing this is because of maintaining support for the old format as well.

In my experience, it would be better to separate the handling of the two different versions.
1 - We probably don't want to liberal in what accept -- which is the current effect
2 - In my experience, counter-intuitively, separating version tends to be cleaner in long term.
3 - Performance will be better with fewer branches

You probably have to send a varied workload to see that difference.  CPU branch predictors are surprisingly good.

Reply:
Yep this is here for backwards compatibility.

Reply:
The way I had imagined this working - and what we implemented and what I think we wanted to evaluate on the Java tracer side - was that all of the string values in the spans would be dictionary encoded, so always integers. To be transparent, I didn't know that the agent currently supported binary values, which is cool but not a feature we'd taken advantage of in the Java tracer. 

I don't think this means we can't evaluate what we wanted to: if we allow the _dictionary_ to be heterogeneously typed, potentially containing any type convertible to a string (msgpack nil, string, bin), we can constrain the keys and values in the messages to _always_ be dictionary codes. This would mean that all of the runtime type/null checking could be done up front as the dictionary is unpacked, and make the bulk of the work - unpacking spans - as simple as possible.

Reply:
This sounds good to me. Let's make sure we don't need `bin` (I'm pretty sure we don't) and go for replacing all strings everywhere with `int` which will be the index of the `string` in the dictionary. I was thinking it would be easiest and nicest implementation and code complexity wise to have only strings in the dictionary. Since `nil` means always empty string, it could be represented as `""` in the dictionary.

If we don't need `bin`, maybe we should remove it. We will need something like that for Method Level Tracing later on but it might be better to be a different type of field.

How does this sound?

Reply:
Update: I left `int` as the only option when parsing strings with the new endpoint (removed `bin` and `string`) and there seems to be a bit of gain:
```
gabriel.aszalos@datadog-gbbr:datadog-agent (gbbr/v0.5)$ benchcmp old.txt no-switch.txt 
benchmark                     old ns/op     new ns/op     delta
BenchmarkDecodeMsgArray-8     1989          1911          -3.92%

benchmark                     old MB/s     new MB/s     speedup
BenchmarkDecodeMsgArray-8     144.26       150.17       1.04x

benchmark                     old allocs     new allocs     delta
BenchmarkDecodeMsgArray-8     14             14             +0.00%

benchmark                     old bytes     new bytes     delta
BenchmarkDecodeMsgArray-8     528           528           +0.00%
```

Reply:
Furthermore, eliminating `nil` values, removing the `for` range and `switch` case gives another improvement:
```
$ benchcmp no-switch.txt no-nil.txt 
benchmark                     old ns/op     new ns/op     delta
BenchmarkDecodeMsgArray-8     1911          1819          -4.81%

benchmark                     old MB/s     new MB/s     speedup
BenchmarkDecodeMsgArray-8     150.17       157.78       1.05x
```

Reply:
I don't think we need to give up `bin` to do this. I think it's just that I never imagined that each key/value could be dictionary encoded or be an inline `string` value. The structure of the dictionary would be as follows

```
[`nil`, `string|bin`, `string|bin`,...]
```

and your unpacking of the dictionary would do the type resolution, e.g. map the byte `0xC0` to whichever default value _makes sense to the trace-agent, as opposed to the tracer_, then based on the type of the value in the dictionary decide whether to call `UnsafeString` `parseString` or whichever the right method is (which you would know better than we do on the Java tracer side) based on whether it's `bin` or `string`. That way you don't lose flexibility to support `bin` and can deduplicate `bin` values too if necessary.

I don't think we'd expect to see huge improvements in performance unless the types were very unpredictable, but it will probably make your inner loop a bit denser, but I think ultimately it's simpler. If we want other tracers to get on board and try to leverage this, we need the codec to be really simple.

Reply:
I just saw your update - it's great that there seem to be some small performance gains.

Reply:
Thanks. I don't have strong feelings around whether to keep `bin` or not. I'd say if it's not used we can remove it for the new endpoint. I suspect some older encoders encoded strings as bytes as this was there to support that.

One question is though: if `int` are replacing all strings, and a `nil` string is known to be equal to `""`, why not have the first element in the array be `""`? This way the dictionary is an array of `[]string` as opposed to array of `string|nil`.

Reply:
How would we differentiate between null and the empty string if we did that? If we encode the first element as `nil` and the empty string as a positive dictionary code, we could decide in a centralised place (the trace agent) how to represent null. We could decide in one place (the dictionary unpacking code) whether to conflate null with the empty string, or to introduce a default value like "na". Does it seem more flexible to give the trace agent control over how to represent null strings? If it's a "deal breaker" we can conflate null with the empty string in the tracer, but I think we lose information this way.

Reply:
I'd be fine with keeping `nil` if that were the case but today there is no such scenario where there is a difference between empty string and nil. The whole map (both metrics and meta) are decoded into type `map[string]string` and `map[string]float64`. Neither of these can take a value of `nil`. Basically `nil` is not a valid value for `string`.

Reply:
I think the problem is assuming that a nil is equal to empty string.  That's not universally true.

I don't think we should change the user experience as part of this change.  So unless we treated nil & empty string as equivalent previously, we shouldn't start in this protocol change.

Reply:
> I don't think we should change the user experience as part of this change. So unless we treated nil & empty string as equivalent previously, we shouldn't start in this protocol change.

I agree. My problem lies solely with putting `nil` in the dictionary. I want the dictionary to be a clear type `[]string`. If anyone wants to send `nil` in place of strings, I'm happy to support that as `nil` is 1 byte in msgpack encoding. What I don't want is the `[]string` dictionary to be a `[](string|nil)` dictionary. I guess what @richardstartin doesn't want is for strings not to be `int|nil` (which they are anyway today).

Am I right?

Reply:
> We were always treating nil as equal to empty string in the whole pipeline starting from the trace-agent, and there is no other way.

We were writing `nil` when strings were null before, and the trace-agent was applying that change; we were never conflating null with the empty string in the Java tracer. Now that we move all string values into the dictionary and out of the spans, that includes null, and I don't think it should change.

Reply:
I'll try to clarify. These are the span fields that we have today along with their types:

* Service   (int)
* Name      (int)
* Resource  (int)
* TraceID   (uint64)
* SpanID    (uint64)
* ParentID  (uint64)
* Start     (int64)
* Duration  (int64)
* Error     (int32)
* Meta      (map[int]int)
* Metrics   (map[int]float64)
* Type      (int)

This is the behaviour expected by the decoder:
* Any of the above fields can be set to `nil` when they are absent, regardless of their type, be it `string`, `int64`, `uint64`, etc.
* When the `Meta` and `Metrics` maps are present, neither their keys nor values can ever be `nil`, only `string`.

What I'm proposing is that `nil` has no place in the dictionary because it is a valid value anywhere above. The dictionary should be string only. I am *NOT* suggesting conflating empty string with `nil`. Sorry if I was unclear about that. 

If we would put `nil` in the dictionary, it would not change the fact that the above fields can still be `nil`, so besides it being confusing and seemingly like a hack, it wouldn't help.

Reply:
I'm resolving this and copying the above comment to the relevant thread. I think we've already clarified the original comment. If not, please feel free to add more Doug/Richard and even re-open.

Reply:
I don't think this is resolved. I quote what I said above:

> The way I had imagined this working - and what we implemented and what I think we wanted to evaluate on the Java tracer side - was that all of the string values in the spans would be dictionary encoded, so always integers. To be transparent, I didn't know that the agent currently supported binary values, which is cool but not a feature we'd taken advantage of in the Java tracer.
>
> I don't think this means we can't evaluate what we wanted to: if we allow the dictionary to be heterogeneously typed, potentially containing any type convertible to a string (msgpack nil, string, bin), we can constrain the keys and values in the messages to always be dictionary codes. This would mean that all of the runtime type/null checking could be done up front as the dictionary is unpacked, and make the bulk of the work - unpacking spans - as simple as possible.

And your immediate response:

> This sounds good to me. Let's make sure we don't need bin (I'm pretty sure we don't) and go for replacing all strings everywhere with int which will be the index of the string in the dictionary. I was thinking the easiest, nicest implementation with less complex code would be to have only strings in the dictionary. Since nil means always empty string, it could be represented as "" in the dictionary.

At this point I thought you got it - the values in the message will be dictionary encoded, and there will be no `nil` values in the message at points you would need to dictionary decode, and I thought we just had to discuss how to represent missing values in the encoding itself; you would prefer the empty string and I prefer `nil`. Had I misunderstood where this conversation had got to?